server/src/main/java/org/opensearch/transport/ClusterConnectionManager.java

```diff
@@ -64,6 +64,15 @@ public class ClusterConnectionManager implements ConnectionManager {
 
     private final ConcurrentMap<DiscoveryNode, Transport.Connection> connectedNodes = ConcurrentCollections.newConcurrentMap();
     private final ConcurrentMap<DiscoveryNode, ListenableFuture<Void>> pendingConnections = ConcurrentCollections.newConcurrentMap();
+    /**
+     This set is used only by cluster-manager nodes.
+     Nodes are marked as pending disconnect right before cluster state publish phase.
+     They are cleared up as part of cluster state apply commit phase
+     This is to avoid connections from being made to nodes that are in the process of leaving the cluster
```

Comment:
We will require tests in `JoinHelperTests` to verify if the followers are able to handle the new error and re-attempt to join the cluster. Can we add them ?

Reply:
`JoinHelperTests` is manually calling `sendJoinRequest` and validating the functionality, it might not be possible to verify the retry logic here since that happens through discovery in `PeerFinder`

`Coordinator.handleJoinRequest` already had previous cases where join request could fail and it would call the `JoinCallback.onFailure`. Even the `connectToNode` call in `handleJoinRequest` already had logic to call the join callback if it failed with any exception.

I believe the ITs we have added are sufficient since we are able to validate that the node is joining back successfully

Reply:
i have added an assertion in NodeJoinLeftIT that validates that the correct exception is showing up when the join request fails. The test itself validates that the join eventually succeeds